### ISSUE TITLE
Chaining more than two dialogs causes serialization to fail

### ISSUE THREAD

**Issue by vossccp** [Microsoft/BotBuilder#1557]
When calling below dialog, I get a ClosureCaptureException. It seems, the chaining creates a intermediate type which is not serializable:

`			var prompts = new[] { "p1", "p2", "p3" };

			var query = from x in new PromptDialog.PromptString(prompts[0], prompts[0], attempts: 1)
						from y in new PromptDialog.PromptString(prompts[1], prompts[1], attempts: 1)
						from z in new PromptDialog.PromptString(prompts[2], prompts[2], attempts: 1)
						select string.Join(" ", x, y, z);

			context.Call(query.PostToUser(), async (dialogContext, awaitableResult) =>
			{
				try
				{
					var result = await awaitableResult;

					if (result != null)
					{
						await dialogContext.PostAsync("You have selected " + result);
					}
				}
				catch (TooManyAttemptsException)
				{
					await dialogContext.PostAsync("Aborting");
				}

				dialogContext.Wait(MessageReceived);
			});
`

**Comment by willportnoy** [comment]
Yes, I'd recommend using the reflection-based serialization surrogate for SelectMany, as we do in our unit test:
https://github.com/Microsoft/BotBuilder/blob/master/CSharp/Tests/Microsoft.Bot.Builder.Tests/ChainTests.cs#L87

The issue is that the anonymous class generated by the compiler is not marked as serializable:
https://github.com/Microsoft/BotBuilder/blob/master/CSharp/Library/Dialogs/Chain.cs#L578

some more information:

https://docs.botframework.com/en-us/technical-faq/#how-can-i-reference-non-serializable-services-from-my-c-dialogs
